core-rust/state-manager/src/state_manager.rs

```diff
@@ -712,123 +711,132 @@ where
+            write_store.deref(),
+            series_executor.epoch_identifiers(),
+            series_executor.latest_state_version(),
+            &prepared_transactions,
+        );
+        if resultant_transaction_root != commit_ledger_header.hashes.transaction_root {
+            warn!(
+                "resultant transaction root at version {} differs from the proof ({} != {})",
+                commit_state_version,
+                resultant_transaction_root,
+                commit_ledger_header.hashes.transaction_root
+            );
+            return Err(InvalidCommitRequestError::TransactionRootMismatch);
+        }
+
         let mut committed_transaction_bundles = Vec::new();
         let mut substate_store_update = SubstateStoreUpdate::new();
         let mut state_tree_update = HashTreeUpdate::new();
-        let transaction_tree_len = AccuTreeEpochHandler::new(
+        let epoch_accu_trees = EpochAwareAccuTreeFactory::new(
             series_executor.epoch_identifiers().state_version,
             series_executor.latest_state_version(),
-        )
-        .current_accu_tree_len();
-        let mut transaction_tree_slice_merger = AccuTreeSliceMerger::new(transaction_tree_len);
-        let mut receipt_tree_slice_merger = AccuTreeSliceMerger::new(transaction_tree_len);
+        );
+        let mut transaction_tree_slice_merger = epoch_accu_trees.create_merger();
+        let mut receipt_tree_slice_merger = epoch_accu_trees.create_merger();
         let mut intent_hashes = Vec::new();
 
         let mut result_receipts = vec![];
-        for (i, (raw, prepared)) in prepared_transactions.into_iter().enumerate() {
+        for (raw, prepared) in commit_request
+            .transactions
+            .into_iter()
+            .zip(prepared_transactions)
+        {
             let (validated, config_type) = if genesis {
                 (
                     self.ledger_transaction_validator.validate_genesis(prepared),
                     ConfigType::Genesis,
                 )
             } else {
                 (
-                    self
-                        .ledger_transaction_validator
+                    self.ledger_transaction_validator
                         .validate_user_or_round_update(prepared)
                         .unwrap_or_else(|error| {
-                            panic!(
-                                "Committed transaction is not valid - likely byzantine quorum: {error:?}"
-                            );
+                            panic!("cannot validate transaction to be committed: {error:?}");
                         }),
                     ConfigType::Regular,
                 )
             };
 
-            let (
-                state_hash_tree_diff,
-                transaction_tree_slice,
-                receipt_tree_slice,
-                local_receipt,
-                database_updates,
-            ) = {
```

Comment:
_(note from author)_

I removed this `let ... = {...}` (and a couple of `.clone()`s) since the entire body got simplified recently and I feel it's readable without the "scoping".
Let me know if you'd rather have this back. But I have plans for some further simplifications here too.

Reply:
This is all good :)